### ISSUE TITLE
Consider shell-agnostic activate.d/deactivate.d mechanism

### ISSUE THREAD

**Issue by kalefranz** [conda/conda#6820]
Currently `activate.d` and `deactivate.d` directories contain files meant to be sourced by the currently-active shell.  E.g. `.sh` files for bash shells; `.bat` files for cmd.exe shells.  In those two files alone there's potential need for duplicated logic.  It's also a common problem that `.sh` files are often naively written for bash, but should in fact be written for posix shell compliance.  These shell-specific files also completely neglect powershell, xonsh, tcsh, fish, and any other future shells we might easily support for conda with a simple shell wrapper addition.

We therefore need a shell-agnostic method for specifying changes to process environment variables during environment activation and deactivation.

**Comment by kalefranz** [comment]
Right now I see three general options

1. Putting .py files in `activate.d`/`deactivate.d` that contain convention-defined function names with convention-defined function signatures that can be called directly within conda's python process.
2. Using subprocess calls to generic executables in `activate.d`/`deactivate.d`, which use one the passed environment variables as inputs, and are expected to output json of some documented form.
3. To hedge against #6819, and the possibility that conda might not have a strict python dependency in the future, a variation option 1 that instead uses lua as an interpreter.  Lua is a <200kB package that has only readline and ncurses as dependencies.  Python is a 27MB package with quite a few more dependencies.  This would mean, however, the lua, along with python, would be hard dependencies of conda.

**Comment by njalerikson** [comment]
Is there ever a use case for these scripts that is activating/deactivating something not related to setting/unsetting environment variables? With the current design more or less anything can be done within those scripts. I can imagine some interesting use cases of activate scripts calling lmod and I do not see a clear way to preserve such information via the first two ideas.

The third idea is interesting. I did just mention lmod so maybe...

**Comment by kalefranz** [comment]
I had to google lmod. https://lmod.readthedocs.io/en/latest/

**Comment by njalerikson** [comment]
After thinking about this and discussing with coworkers I think you should consider the option to integrate LMOD for the use of these `{de,}activate.d` scripts.

Just out of the box it would support a good number of shells, quoting the [TACC page](https://www.tacc.utexas.edu/research-development/tacc-projects/lmod):
> All the popular shells are supported: bash, ksh, csh, tcsh, zsh. LMOD is also available for perl and python.

This list of supported shells is clearly missing Windows support but since Lua is supported on Windows it would just be a matter of investing the time to add Windows abilities to LMOD.

And then you could leave the option to create shell specific `{de,}activate.d}` scripts for the more exotic shells (e.g. xonsh, fish). But I'd assume that even these can be added.

Not only would this be a reasonably light weight way to address these scripts but LMOD could even be used for the conda activate/deactivate processes.

**Comment by msarahan** [comment]
LMOD does seem to be a natural choice to me.  Conda-build could specify some syntax in meta.yaml that would output lua scripts for LMOD, and shell-specific scripts as necessary.

```
build:
  env_vars:
    SOMEVAR: value
```

We'd need to differentiate setting vars from adding to vars somehow.  This wouldn't necessarily be required - people could create these scripts manually themselves without using conda-build.  I think using conda-build might make the transition easier, though.

**Comment by kalefranz** [comment]
https://github.com/conda/conda/issues/6945#issuecomment-378842065

> Having an activate verbosity feature that the scripts can query would be great though.

**Comment by kalefranz** [comment]
xref: https://github.com/conda/conda/issues/1037

**Comment by jakirkham** [comment]
Probably could learn some things from [`direnv`]( https://github.com/direnv/direnv ) here.

**Comment by tkf** [comment]
As a low-level API, is it possible to add a conda command which prints a (say) JSON containing all the environment variables (and other information for activation if any)?

We need it for conda/python integration in julia https://github.com/JuliaPy/Conda.jl/issues/125#issuecomment-423771056. I know that `conda run` is coming back but it still lacks Windows support #7320. Furthermore, `conda run` is not enough since we need to update DLL search path within Julia process.

**Comment by LiberalArtist** [comment]
Re @tkf's point, as a lighter-weight alternative to JSON, there is already a trivially parsable format for environment variables used by GNU versions of `/usr/bin/env`: the command `/usr/bin/env --null` writes all environment variables to standard output as key–value pairs, with `=` used to delimit the end of the name and the null character used to delimit the end of the value. This is unambiguous, since `=` is not allowed in names (on either [Windows](https://docs.microsoft.com/en-us/windows/desktop/ProcThread/environment-variables) or [Unix-like systems](http://pubs.opengroup.org/onlinepubs. /000095399/basedefs/xbd_chap08.html)) and null is not allowed in values. (Unfortunately, non-GNU versions of `/usr/bin/env`, e.g. on Mac, don't seem to support this option.)

**Comment by dhirschfeld** [comment]
> *Currently activate.d and deactivate.d directories contain files meant to be sourced by the currently-active shell*

I'll just mention that this doesn't seem to be happening for powershell: https://github.com/conda/conda/issues/8101

**Comment by cgranade** [comment]
@LiberalArtist: Apologies for pinging a pretty long-dormant thread, but the main issue I could see with a `NAME=value` format would be adding a folder to `$PATH` without modifying anything else that's already there.

**Comment by BrenBarn** [comment]
This seems like an important issue for getting conda run to work properly (#2379).

Was it a deliberate design decision for conda to allow running arbitrary code in these activate.d scripts?  Is there any information about how common it is for packages to use the scripts to do anything besides setting (or maybe removing or otherwise interacting with) environment variables?  Is there a possibility to reduce the "degrees of freedom" here by introducing a declarative format for specifying environment changes, so that packages just declare what env vars they want (akin to [Jupyter kernelspecs](https://jupyter-client.readthedocs.io/en/stable/kernels.html#kernel-specs))?  In the long it seems it would simplify matters greatly if we just moved away from conda (and the package scripts it supports) doing its setup by running shell commands.

**Comment by astrojuanlu** [comment]
Yes, I can still reproduce the issue in the latest version of Conda... Like most of the other issues closed by the naughty bot.

**Comment by BrenBarn** [comment]
Agreed, the issue is clearly not resolved as long as `activate.d` scripts still exist. . . :-)

**Comment by jakirkham** [comment]
Maybe another avenue is looking at [Cobra's shell completion]( https://github.com/spf13/cobra/blob/master/shell_completions.md )

(Thanks @mariusvniekerk for the suggestion)

**Comment by wolfv** [comment]
One approach I tried in the past was to work on a `multisheller` python tool that takes a pythonic DSL and translates that into bash/zsh/powershell/fish...

We could then ship multisheller files as part of the packages and integrate that tool with conda/mamba, or translate the multisheller files at build time into the different activation scripts. The upside is that some more complex things like `if / else` and adding / removing to `$PATH` work, which is not possible with just environment variables.

It hasn't taken off yet, but there are some people using it and it works :)

**Comment by dhirschfeld** [comment]
> It hasn't taken off yet, but there are some people using it and it works

I've looked into it as I think there's a lot of value in it. I think a lot of packages with activation scripts probably don't work properly in all shells so having a tool which makes it easier to support multiple shells will make the `conda`/`mamaba` ecosystem much better for anyone not using `bash`.

**Comment by cmarqu** [comment]
@wolfv So kind of https://github.com/cea-hpc/modules implemented in Python? Maybe it could be compatible with (a subset of) the modulefile syntax which is pretty widespread in HPC and chip design environments.

**Comment by dhirschfeld** [comment]
I'd actually go further than just allowing `multisheller` scripts. IME there are two very common cases where a package might use `{de}activate` scripts:
1. Adding/removing environment variables when an environment is activated/deactivated
2. Adding/removing path entries when an environment is activated/deactivated

It might be even nicer to provide a mechanism in the `recipe.yaml` itself to support these use-cases - e.g.
```yaml
env:
  - name: MY_CUSTOM_VAR
    value: "foobar"
path:
  - ${PREFIX}/mypackage/bin
```

The build tool would then generate the appropriate `multisheller` scripts so that the environment variable and path were correctly updated irrespective of the shell you used.

**Comment by wolfv** [comment]
Nope, they are apparently not correctly unset upon deactivation :) (see https://github.com/conda/conda/issues/11726)

We're implementing this now for `micromamba` and maybe we'll fix it already.

**Comment by mfisher87** [comment]
Migrating from an `activate.d` script to JSON in `env_vars.d` seems like a great solution to a problem one of my users encountered.

I searched for documentation of `env_vars.d` and [searched the conda-forge org](https://github.com/search?q=org%3Aconda-forge%20env_vars.d&type=code) for examples of others using it, but came up with nothing. I found some conversational references to it (this issue, and a Stack Overflow post), but that's it. I am hesitating to be (among?) the first maintainer in Conda Forge to adopt this for the Quarto feedstock.

Is `env_vars.d` a supported interface that's not going away? If so, I'll open a PR to document it. Perhaps the first step is to open an issue "document the `env_vars.d` interface" and see what response I get.
